lib/ansible/galaxy/collection/gpg.py

```diff
@@ -0,0 +1,298 @@
+
+import os
+import subprocess
+import textwrap
+
+from dataclasses import dataclass, fields as dc_fields
+from urllib.error import HTTPError, URLError
+
+
+def run_gpg_verify(b_manifest_file, signature, keyring, display):
+    status_fd_read, status_fd_write = os.pipe()
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        b_manifest_file,
+    ]
+    cmd_str = ' '.join(to_text(arg) for arg in cmd)
+    display.vvvv(f"Running command '{cmd}'")
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+        )
+    except (FileNotFoundError, subprocess.SubprocessError) as err:
+        raise AnsibleError(
+            f"Failed during GnuPG verification with command '{cmd_str}': {err}"
+        ) from err
+    else:
+        stdout, stderr = p.communicate(input=signature)
+    finally:
+        os.close(status_fd_write)
+
+    with os.fdopen(status_fd_read) as f:
+        stdout = f.read()
+        display.vvvv(f"{stdout} (exit code {p.returncode})")
+        return stdout, p.returncode
+
+
+def parse_gpg_errors(status_out):
+    for line in status_out.splitlines():
+        if not line:
+            continue
+        try:
+            _dummy, status, remainder = line.split(maxsplit=2)
+        except ValueError:
+            _dummy, status = line.split(maxsplit=1)
+            remainder = None
+        try:
+            cls = GPG_ERROR_MAP[status]
+        except KeyError:
+            continue
+
+        fields = [status]
+        if remainder:
+            fields.extend(
+                remainder.split(
+                    None,
+                    len(dc_fields(cls)) - 2
+                )
+            )
+
+        yield cls(*fields)
+
+
+def get_signature_from_source(source, display=None):
+    if display is not None:
+        display.vvvv(f"Using signature at {source}")
+    try:
+        with open_url(
+            to_text(source),
+            http_agent=user_agent(),
+            follow_redirects='safe'
+        ) as resp:
+            return resp.read()
+    except (HTTPError, URLError) as e:
+        raise AnsibleError(
+            f"Failed to get signature for collection verification from '{source}': {e}"
+        ) from e
+
+
+# TODO: Optimize with slots=True when the min Python version for the controller is >= 3.10
+@dataclass(frozen=True)
+class GpgBaseError(Exception):
+    status: str
+
+    def __post_init__(self):
+        for field in dc_fields(self):
+            super().__setattr__(field.name, field.type(getattr(self, field.name)))
+
+
+@dataclass(frozen=True)
+class GpgExpSig(GpgBaseError):
+    keyid: str
+    username: str
+
+    @staticmethod
+    def explain():
```

Comment:
Two points here.
1) Since the class already uses some variable type annotations, it would only be consistent to have function annotations on its methods too:
```suggestion
    def explain() -> str:
```

Reply:
2) Bikeshed the name/implementation? `explain` is ambiguous, it doesn't really explain what it explains and if you look into the method body, you realize that it doesn't explain anything but rather returns an explanation string. When I see `explain()`, I expect some action to be happening, maybe some UI boxes show up or messages printed out on screen. But this just returns a string.

Reply:
I just realized that there's also the third thing.
Looking into this more, I'd say that the implementation could be changed by adding
```python
@classmethod
def get_gpg_error_description(cls) -> str:
    """Return the current class description."""
    return cls.__doc__
```
to the base class `GpgBaseError` and moving this string to into class docstrings.
So you won't need to implement an extra method in each child class while having a nice REPL-introspectable docstring in them too.

Reply:
Good idea. Implemented https://github.com/ansible/ansible/pull/76681#discussion_r785211797.